lib/logstash/outputs/s3.rb

```diff
@@ -73,285 +61,354 @@
+  config :temporary_directory, :validate => :string, :default => File.join(Dir.tmpdir, "logstash")
+
+  # Specify a prefix to the uploaded filename, this can simulate directories on S3
+  config :prefix, :validate => :string, :default => ''
+
+  # Specify how many workers to use to upload the files to S3
+  config :upload_workers_count, :validate => :number, :default => 1
+
+  # Exposed attributes for testing purpose.
+  attr_accessor :tempfile
+  attr_reader :page_counter
+  attr_reader :s3
+
+  def aws_s3_config
+    @logger.info("Registering s3 output", :bucket => @bucket, :endpoint_region => @region)
+    @s3 = AWS::S3.new(aws_options_hash)
+  end
 
- # Aws access_key.
- config :access_key_id, :validate => :string
+  def aws_service_endpoint(region)
+    # Make the deprecated endpoint_region work
+    # TODO: (ph) Remove this after deprecation.
+    if @endpoint_region
+      region_to_use = @endpoint_region
+    else
+      region_to_use = region
+    end
 
- # Aws secret_access_key
- config :secret_access_key, :validate => :string
+    return {
+      :s3_endpoint => region_to_use == 'us-east-1' ? 's3.amazonaws.com' : "s3-#{region_to_use}.amazonaws.com"
+    }
+  end
 
- # S3 bucket
- config :bucket, :validate => :string
+  public
+  def write_on_bucket(file)
+    # find and use the bucket
+    bucket = @s3.buckets[@bucket]
 
- # Aws endpoint_region
- config :endpoint_region, :validate => ["us-east-1", "us-west-1", "us-west-2",
-                                        "eu-west-1", "ap-southeast-1", "ap-southeast-2",
-                                        "ap-northeast-1", "sa-east-1", "us-gov-west-1"], :default => "us-east-1"
+    remote_filename = "#{@prefix}#{File.basename(file)}"
 
- # Set the size of file in KB, this means that files on bucket when have dimension > file_size, they are stored in two or more file.
- # If you have tags then it will generate a specific size file for every tags
- ##NOTE: define size of file is the better thing, because generate a local temporary file on disk and then put it in bucket.
- config :size_file, :validate => :number, :default => 0
+    @logger.debug("S3: ready to write file in bucket", :remote_filename => remote_filename, :bucket => @bucket)
 
- # Set the time, in minutes, to close the current sub_time_section of bucket.
- # If you define file_size you have a number of files in consideration of the section and the current tag.
- # 0 stay all time on listerner, beware if you specific 0 and size_file 0, because you will not put the file on bucket,
- # for now the only thing this plugin can do is to put the file when logstash restart.
- config :time_file, :validate => :number, :default => 0
+    begin
+      # prepare for write the file
+      object = bucket.objects[remote_filename]
+      object.write(:file => file, :acl => @canned_acl)
+    rescue AWS::Errors::Base => e
+      @logger.error("S3: AWS error", :error => e)
+      raise LogStash::ConfigurationError, "AWS Configuration Error"
```

Comment:
Not sure if ConfigurationError is the right kind of exception to raise when a write operation fails...maybe LogStash::Error?